File: Python3/Jeu_min_max.py


# Importations
import os
import random
import re

# Données
CHEMIN_DOSSIER = os.path.dirname(os.path.abspath(__file__))
FICHIER_OPTIONS = os.path.join(CHEMIN_DOSSIER, "minmax_options.txt")

# Fonctions utilitaires
## Lecture d'un argument de commande

"""
	Lecture sécurisée d'un entier depuis l'utilisateur
	Contrôle l'appartenance à un intervalle [minimum, maximum]
	@message message à afficher
	@minimum borne inférieure
	@maximum borne supérieure
"""
def lire_entier(message, minimum, maximum):
	while True:
		print(message, end='')
		entree = input().strip()
		if re.fullmatch(r"\d+", entree) and minimum <= int(entree) <= maximum:
			return int(entree)
		print(f"\nVeuillez entrer un entier entre {minimum} et {maximum}.")

# Fonctions utilitaires principales
## Création d'une tâche

"""
	Charge les options sauvegardées depuis le fichier
	Retourne une liste de trois entiers [minimum, maximum, nombre_tours]
	@return liste d'options
"""
def charger_options():
	if os.path.exists(FICHIER_OPTIONS):
		with open(FICHIER_OPTIONS, "r", encoding="utf-8") as fichier_options:
			donnees = fichier_options.read().split(",")
		options = [int(valeur) for valeur in donnees if valeur.isdigit()]
		if len(options) == 3:
			return options
	return [1, 100, 5]

"""
	Sauvegarde les options (minimum, maximum, nombre_tours) dans le fichier
	Remplace l'ancien contenu
	@minimum borne inférieure
	@maximum borne supérieure
	@nombre_tours nombre de tours
"""
def sauvegarder_options(minimum, maximum, nombre_tours):
	with open(FICHIER_OPTIONS, "w", encoding="utf-8") as fichier_options:
		fichier_options.write(f"{minimum},{maximum},{nombre_tours}")

# Fonctions principales

"""
	Affiche le menu principal du programme
"""
def afficher_menu_principal():
	print("\n#### Menu ####\n1 : Jouer\n2 : Options\n3 : Quitter")

"""
	Affiche le menu d'options avec valeurs courantes
	@minimum borne inférieure
	@maximum borne supérieure
	@nombre_tours nombre de tours
"""
def afficher_menu_options(minimum, maximum, nombre_tours):
	print(f"\n#### Menu options ####\n1 : Choisir les limites (actuellement : {minimum} - {maximum})\n2 : Nombre de tours max (actuellement : {nombre_tours})")

"""
	Joue une partie, affiche + ou - selon la réponse
	@minimum borne inférieure
	@maximum borne supérieure
	@nombre_tours nombre de tours
"""
def jouer_partie(minimum, maximum, nombre_tours):
	cible = random.randint(minimum, maximum)
	print(f"\nTrouvez entre {minimum} et {maximum} en {nombre_tours} tours")
	for numero_essai in range(1, nombre_tours + 1):
		choix = lire_entier(f"Tour {numero_essai} : ", minimum, maximum)
		if choix == cible:
			print("\nGagné !")
			return
		elif choix > cible:
			print("-")
		else:
			print("+")
	print(f"\nPerdu ! La réppnse était {cible}.")

"""
	Gère le menu des options et sauvegarde si modifiées
	@minimum borne inférieure
	@maximum borne supérieure
	@nombre_tours nombre de tours
	@return tuple (minimum, maximum, nombre_tours)
"""
def gerer_options(minimum, maximum, nombre_tours):
	afficher_menu_options(minimum, maximum, nombre_tours)
	choix_options = lire_entier("? = ", 1, 2)
	if choix_options == 1:
		minimum = lire_entier("\nMin = ", 1, maximum)
		maximum = lire_entier("\nMax = ", minimum, 10000)
	elif choix_options == 2:
		nombre_tours = lire_entier("\nTours max = ", 1, 100)
	sauvegarder_options(minimum, maximum, nombre_tours)
	return minimum, maximum, nombre_tours

# Main

"""
	Boucle principale du programme
	Gère la navigation entre menu, jeu et options
	@none aucun
"""
def main():
	minimum, maximum, nombre_tours = charger_options()
	while True:
		afficher_menu_principal()
		choix_menu = lire_entier("? = ", 1, 3)
		if choix_menu == 1:
			jouer_partie(minimum, maximum, nombre_tours)
		elif choix_menu == 2:
			minimum, maximum, nombre_tours = gerer_options(minimum, maximum, nombre_tours)
		elif choix_menu == 3:
			break

# Lancement du programme
main()

File: Rust/src/Jeu_min_max.rs
//! # Jeu MinMax
//! Dépendance externe requise : rand = "0.8"

// # Importations
use std::fs;
use std::io::{self, Write};
use std::path::PathBuf;
use rand::Rng;

// # Données

// # Fonctions utilitaires
// ## Structure de données

/// Structure des paramètres du jeu
/// 
/// - `min` : borne inférieure
/// - `max` : borne supérieure
/// - `tours` : nombre d’essais
struct Options {
	min: i32,
	max: i32,
	tours: i32,
}

// ## Chemin du fichier d’options

/// Retourne le chemin absolu vers « ./data/minmax_options.txt »
/// Crée le dossier ./data s’il n’existe pas
fn chemin_options() -> PathBuf {
	let mut chemin = std::env::current_dir().expect("Impossible d’obtenir le dossier courant");
	chemin.push("data");
	fs::create_dir_all(&chemin).expect("Impossible de créer ./data");
	chemin.push("minmax_options.txt");
	chemin
}

// ## Lecture d’un entier utilisateur sécurisé

/// Boucle jusqu’à obtenir un entier compris dans l’intervalle donné
///
/// @message Message affiché à l’utilisateur
/// @borne_min Borne minimale autorisée
/// @borne_max Borne maximale autorisée
fn lire_entier(message: &str, borne_min: i32, borne_max: i32) -> i32 {
	loop {
		print!("{message}");
		io::stdout().flush().unwrap();
		let mut entree = String::new(); 
		io::stdin().read_line(&mut entree).unwrap();
		let trimmed = entree.trim();
		if !trimmed.is_empty() && trimmed.chars().all(|c| c.is_ascii_digit()) {
			let valeur = trimmed.parse::<i32>().unwrap();
			if (borne_min..=borne_max).contains(&valeur) {
				return valeur;
			}
		}
		println!("\nVeuillez entrer un entier entre {borne_min} et {borne_max}.");
	}
}

// ## Chargement des options sauvegardées

/// Lit les options depuis le fichier, ou renvoie des valeurs par défaut
fn charger_options() -> Options {
	let chemin = chemin_options();
	if let Ok(contenu) = fs::read_to_string(chemin) {
		let morceaux: Vec<i32> = contenu.split(',').filter_map(|s| s.trim().parse().ok()).collect();
		if morceaux.len() == 3 {
			return Options { min: morceaux[0], max: morceaux[1], tours: morceaux[2] };
		}
	}
	Options { min: 1, max: 100, tours: 5 }
}

// ## Sauvegarde des options

/// Écrit les paramètres actuels dans le fichier d’options
///
/// @options Référence à la structure contenant min, max, tours
fn sauvegarder_options(options: &Options) {
	let chemin = chemin_options();
	let contenu = format!("{},{},{}", options.min, options.max, options.tours);
	let _ = fs::write(chemin, contenu);
}

// # Fonctions principales
// ## Affiche le menu principal

/// Affiche les choix disponibles à l’utilisateur
fn afficher_menu_principal() {
	println!("\n#### Menu ####\n1 : Jouer\n2 : Options\n3 : Quitter");
}

// ## Affiche le menu d’options

/// Affiche les paramètres actuels dans le menu d’options
///
/// @options Référence à la structure de configuration
fn afficher_menu_options(options: &Options) {
	println!(
		"\n#### Menu options ####\n1 : Choisir les limites (actuellement : {} - {})\n2 : Nombre de tours max (actuellement : {})",
		options.min, options.max, options.tours
	);
}

// ## Joue une partie (+ / -)

/// Lance une partie avec un nombre d’essais limité
///
/// Affiche « + » ou « - » selon la comparaison avec la cible.
/// @options Référence aux limites et nombre de tours
fn jouer_partie(options: &Options) {
	let cible = rand::thread_rng().gen_range(options.min..=options.max);
	println!("\nTrouvez entre {} et {} en {} tours", options.min, options.max, options.tours);
	for tentative in 1..=options.tours {
		let choix = lire_entier(&format!("Tour {} : ", tentative), options.min, options.max);
		if choix == cible {
			println!("\nGagné !");
			return;
		}
		println!("{}", if choix > cible { "-" } else { "+" });
	}
	println!("\nPerdu ! La réppnse était {}.", cible);
}

// ## Gère les options et sauvegarde si modifiées

/// Modifie les paramètres selon l’utilisateur et les sauvegarde
///
/// @options Référence mutable à la configuration actuelle
fn gerer_options(options: &mut Options) {
	afficher_menu_options(options);
	let choix_options = lire_entier("? = ", 1, 2);
	if choix_options == 1 {
		options.min = lire_entier("\nMin = ", 1, options.max);
		options.max = lire_entier("\nMax = ", options.min, 10_000);
	} else {
		options.tours = lire_entier("\nTours max = ", 1, 100);
	}
	sauvegarder_options(options);
}

// # Main

/// Point d’entrée principal du programme
fn main() {
	let mut options = charger_options();
	loop {
		afficher_menu_principal();
		match lire_entier("? = ", 1, 3) {
			1 => jouer_partie(&options),
			2 => gerer_options(&mut options),
			3 => break,
			_ => {}
		}
	}
}
